pkg/netpol/eval/internal/k8s/adminnetpol.go

```diff
@@ -0,0 +1,297 @@
+		// AdminNetworkPolicySubject should contain exactly one field
+		// (https://github.com/kubernetes-sigs/network-policy-api/blob/v0.1.5/apis/v1alpha1/shared_types.go#L27))
+		return false, anp.anpErr(netpolerrors.SubjectErrTitle, netpolerrors.SubjectFieldsErr)
+	}
+	if anpSubject.Namespaces != nil {
+		return anp.doesNamespacesFieldMatchPeer(anpSubject.Namespaces, p)
+	}
+	// else: Subject.Pods is not empty (Subject.Pods contains both NamespaceSelector and PodSelector)
+	return anp.doesPodsFieldMatchPeer(anpSubject.Pods, p)
+}
+
+// onlyOnePeersFieldIsSet returns if exactly one of the given objects is set (not nil)
+func onlyOnePeersFieldIsSet(namespacesField *metav1.LabelSelector, podsField *apisv1a.NamespacedPod) bool {
+	return (namespacesField == nil) != (podsField == nil)
+}
+
+// doesNamespacesFieldMatchPeer returns if the given namespaces LabelSelector matches the given peer
+func (anp *AdminNetworkPolicy) doesNamespacesFieldMatchPeer(namespaces *metav1.LabelSelector, peer Peer) (bool, error) {
+	if peer.PeerType() == IPBlockType {
+		return false, nil // namespaces does not select IPs
+	}
+	namespacesSelector, err := anp.parseAdminNetpolLabelSelector(namespaces)
+	if err != nil {
+		return false, err
+	}
+	return namespacesSelector.Matches(labels.Set(peer.GetPeerNamespace().Labels)), nil
+}
+
+// doesPodsFieldMatchPeer returns if the given NamespacedPod object matches the given peer
+// a NamespacedPod object contains both NamespaceSelector and PodSelector
+func (anp *AdminNetworkPolicy) doesPodsFieldMatchPeer(pods *apisv1a.NamespacedPod, peer Peer) (bool, error) {
+	if peer.PeerType() == IPBlockType {
+		return false, nil // pods does not select IPs
+	}
+	nsSelector, err := anp.parseAdminNetpolLabelSelector(&pods.NamespaceSelector)
+	if err != nil {
+		return false, err
+	}
+	podSelector, err := anp.parseAdminNetpolLabelSelector(&pods.PodSelector)
+	if err != nil {
+		return false, err
+	}
+	return nsSelector.Matches(labels.Set(peer.GetPeerNamespace().Labels)) && podSelector.Matches(labels.Set(peer.GetPeerPod().Labels)), nil
+}
+
+// parseAdminNetpolLabelSelector returns a selector of type labels.selector from a LabelSelector in the policy.
+// an error with the admin network policy details returned if fails to convert the selector.
+func (anp *AdminNetworkPolicy) parseAdminNetpolLabelSelector(selector *metav1.LabelSelector) (labels.Selector, error) {
+	selectorRes, err := metav1.LabelSelectorAsSelector(selector)
+	if err != nil {
+		return nil, anp.anpErr(netpolerrors.SelectorErrTitle, err.Error())
+	}
+	return selectorRes, nil
+}
+
+// anpErr returns string format of an admin netpol err
+func (anp *AdminNetworkPolicy) anpErr(title, description string) error {
+	return fmt.Errorf("admin network policy %q %s: %s", anp.Name, title, description)
+}
+
+// GetIngressPolicyConns returns the connections from the ingress rules selecting the src in spec of the adminNetworkPolicy
+//
+//nolint:dupl // this loops Ingress spec - different types
```

Comment:
I am trying to see if generics help with this and more funcs in #403

Reply:
currently tried to see if generics helps to avoid code duplication with  `ingressRuleSelectsPeer` and `egressRuleSelectsPeer` func, but unfortunately it didn't work because of a Go restriction on generic types. 

> "The Go compiler does not support accessing a struct field x.f where x is of type parameter type even if all types in the type parameter’s type set have a field f. We may remove this restriction in a future release."

(taken from https://tip.golang.org/doc/go1.18#generics and still relevant to go 1.21)

i have committed [here](https://github.com/np-guard/netpol-analyzer/pull/403/commits/c1bcdcd99787f151c337859462855a9ee1f54065) (in #403) an explanation "why could not use generics in this case)
and still trying to see if there are other cases where generics can be used